File: 0841-keys-and-rooms/0841-keys-and-rooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
var canVisitAllRooms = function(rooms) {
    const n = rooms.length
    const set = new Set()
    dfs(0)
    return set.size === n
    
    function dfs(room) {
        if (set.has(room)) return
        set.add(room)
        for (const next of rooms[room]) {
            dfs(next)
        }
    }
};

File: 0841-keys-and-rooms/0841-keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = set()
        
        def dfs(room: int) -> None:
            visited.add(room)
            for next_room in rooms[room]:
                if next_room not in visited:
                    dfs(next_room)
        
        dfs(0)        
        return len(visited) == len(rooms)
        
        #
        
        n = len(rooms)
        visited = [0] * n
        q = deque([0])
        while q:
            room = q.popleft()
            visited[room] = 1
            for next_room in rooms[room]:
                if visited[next_room] != 1:
                    q.append(next_room)
        return sum(visited) == n